dev-tools/mage/integtest.go

```diff
@@ -0,0 +1,236 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+)
+
+const (
+	// BEATS_DOCKER_INTEG_TEST_ENV is used to indicate that we are inside of the
+	// Docker integration test environment (e.g. in a container).
+	beatsDockerIntegTestEnvVar = "BEATS_DOCKER_INTEG_TEST_ENV"
+)
+
+var (
+	integTestUseCount     int32      // Reference count for the integ test env.
+	integTestUseCountLock sync.Mutex // Lock to guard integTestUseCount.
+
+	integTestLock sync.Mutex // Only allow one integration test at a time.
+)
+
+// Integration Test Configuration
+var (
+	// StackEnvironment specifies what testing environment
+	// to use (like snapshot (default), latest, 5x). Formerly known as
+	// TESTING_ENVIRONMENT.
+	StackEnvironment = EnvOr("STACK_ENVIRONMENT", "snapshot")
+)
+
+// AddIntegTestUsage increments the use count for the integration test
+// environment and prevents it from being stopped until the last call to
+// StopIntegTestEnv(). You should also pair this with
+// 'defer StopIntegTestEnv()'.
+//
+// This allows for the same environment to be reused by multiple tests (like
+// both Go and Python) without tearing it down in between runs.
+func AddIntegTestUsage() {
+	if IsInIntegTestEnv() {
+		return
+	}
+
+	integTestUseCountLock.Lock()
+	defer integTestUseCountLock.Unlock()
+	integTestUseCount++
+}
+
+// StopIntegTestEnv will stop and removing the integration test environment
+// (e.g. docker-compose rm --stop --force) when there are no more users
+// of the environment.
+func StopIntegTestEnv() error {
+	if IsInIntegTestEnv() {
+		return nil
+	}
+
+	integTestUseCountLock.Lock()
+	defer integTestUseCountLock.Unlock()
+	if integTestUseCount == 0 {
+		panic("integTestUseCount is 0. Did you call AddIntegTestUsage()?")
+	}
+
+	integTestUseCount--
+	if integTestUseCount > 0 {
+		return nil
+	}
+
+	if err := haveIntegTestEnvRequirements(); err != nil {
```

Comment:
I think we should avoid _silent errors_ and at least log the content of the error before returning

Reply:
I will update the comment. The reason for ignoring the error is that the error will already have been encounter by the user when `RunIntegTest` was executed. So this check avoid double logging.